### PR TITLE
:sparkles: [FEAT] Apple 로그인 v1 구현

### PR DESCRIPTION
**Pull request by 5hseok**
## #️⃣ Related Issue

Linked Feature backlog: #9 

## 📝 Purpose of PR

Apple Login 구현

## Contents

로컬에서는 테스트가 되지 않는 Apple OAuth입니다. 
이 코드가 배포되면 웹에서 먼저 동작을 보며 코드 수정이 시행될 예정입니다(앱과 연동 시 로직 상 오류가 아니라 설정 오류만 발생하도록)

**로직**
1. ios에서 애플 서버로 요청을 보내 인가 코드를 얻어옵니다.
2. 받은 인가 코드와 기기의 deviceToken을 서버의 /api/auth/login/apple로 POST합니다.
3. 서버는 이를 받아 Client_Secret을 만들고 인가 코드와 client_secret을 이용하여 사용자의 정보를 애플 서버에서 가져옵니다.
4. 이 정보를 디코딩하여 이메일, 고유값을 얻어낸다면 DB에 저장하고 서버 고유의 jwtToken을 생성해 ios에 전달합니다.

## Checklist

-   [ ] Does commit messages follow the convention?
-   [ ] Are variable names brief but descriptive?
-   [ ] Is the code clean and easy to understand?
-   [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

온보딩 작업 혜주님께서 진행 중이시라면 swagger나 notion에 명세 올려두는 작업 진행하겠습니다. 
아직 온보딩 부분 작업 시작 안하셨다면 제가 테이블 정의해서 시작하겠습니다. 
@sueelly

### CODE REVIEW

File: src/main/java/ac/dnd/dodal/common/util/OAuth2Util.java [hunk 1]
```diff
@@ -47,4 +84,74 @@ public KakaoUserInfoDto getKakaoUserInfo(String accessToken) {
...
+
+      KeyFactory keyFactory = KeyFactory.getInstance("EC");
+      return keyFactory.generatePrivate(keySpec);
+
+    } catch (Exception e) {
+      throw new RuntimeException("Error converting private key from String", e);
+    }
+  }
+
+  public JsonElement verifyAuthorizationCode(String code) {
```
Comment (5hseok) [inline src/main/java/ac/dnd/dodal/common/util/OAuth2Util.java:122]: 여기서 토큰을 검증하여 사용자 정보를 담은 토큰을 가져옵니다.

File: src/main/java/ac/dnd/dodal/common/util/OAuth2Util.java [hunk 2]
```diff
@@ -47,4 +84,74 @@ public KakaoUserInfoDto getKakaoUserInfo(String accessToken) {
...
+            String.class);
+
+    if (response.getBody() == null) {
+      throw new RuntimeException("Apple API 요청에 실패했습니다.");
+    }
+
+      return JsonParser.parseString(response.getBody());
+  }
+
+  public <T> T decodePayload(String token, Class<T> targetClass) {
```
Comment (5hseok) [inline src/main/java/ac/dnd/dodal/common/util/OAuth2Util.java:143]: 여기서는 사용자의 정보를 담은 jwt 토큰을 디코딩하여 서비스에 필요한 이메일, sub값 등을 가져오는 메서드입니다.

File: src/main/java/ac/dnd/dodal/domain/user/model/User.java [hunk 3]
```diff
@@ -60,6 +63,10 @@ public void updateUserNickname(String nickname) {
         }
     }
 
+    public void updateAppleRefreshToken(String appleRefreshToken) {
```
Comment (5hseok) [inline src/main/java/ac/dnd/dodal/domain/user/model/User.java:66]: 애플 서버에서 사용할 refreshToken은 최초 로그인 시 한번만 제공하기 때문에 따로 저장하는 메서드와 필드를 추가했습니다

File: src/main/resources/db/migration/V4__add_apple_refresh_token.sql
Comment (sueelly) [inline src/main/resources/db/migration/V4__add_apple_refresh_token.sql:1]: 앞으로 flyway는 서로 어디까지 썼는지 버전을 공지해줘야할것같아요!
현재 6까지 있습니다. 그리고 apple refresh token을 저장해야할 필요성을 잘 모르겠어요

File: src/main/java/ac/dnd/dodal/domain/user/model/User.java (hunk 3 above)
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/user/model/User.java:66]: apple의 refreshtoken을 사용하게 되는 경우가 다시 apple에 접근할 때 말고는 없을 듯 한데, 해당 케이스는 오직 탈퇴할 때밖에 없는 것 같아요. 추후에 확장될 케이스도 보이지 않고요. 현재 로직에서 사용되는 부분도 없고요
탈퇴할 때는 본인인증 겸 다시 로그인해서 받아오는 게 맞다고 생각하기도 해서 저장은 하지 않는 게 어떨까요?

**Review by sueelly — CHANGES_REQUESTED** [review]

File: src/main/java/ac/dnd/dodal/domain/user/model/User.java (hunk 3 above)
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/domain/user/model/User.java:66]: 카카오의 accesstoken과 같다고 보시면 될 것 같아요

**Comment by sueelly** [comment]
온보딩은 아직입니다

File: src/main/java/ac/dnd/dodal/domain/user/model/User.java (hunk 3 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/domain/user/model/User.java:66]: 최초 1회만 제공된다길래 모든 로그인 요청에 대해서 최초 1회만 가능할 걸로 오해하고 있었네요...ㅎ
저희 서버는 자체적인 JWT로 운영되니까 refreshToken을 사용해서 accessToken을 갱신할 때 refreshToken도 갱신되게(자동 로그인) 추후에 추가하면 될 것 같고, 애플 서버에서는 사용자 정보를 가져오기 위해 사용하는 것이니까 말씀해주신 것처럼 굳이 apple refreshToken을 저장하지 않고 탈퇴 시에 다시 로그인 시키는 것이 좋을 것 같네요!!
다시 수정해서 올려두겠습니다.

**Review by sueelly — APPROVED** [review]
